src/compiler/checker.ts

```diff
@@ -1835,6 +1838,23 @@ export function createTypeChecker(host: TypeCheckerHost): TypeChecker {
         typeHasCallOrConstructSignatures,
     };
 
+    function getContextualStringLiteralCompletionTypes(expression: Expression) {
+        const seen = new Map<string, true>();
+
+        return [
+            ...getStringLiteralTypes(getContextualTypeFromParent(expression, checker, ContextFlags.None), seen),
+            ...getStringLiteralTypes(getContextualTypeFromParent(expression, checker, ContextFlags.Completions), seen)
+        ]
+    }
+
+    function getStringLiteralTypes(type: Type | undefined, uniques = new Map<string, true>()): readonly StringLiteralType[] {

```

Comment:
This is a copy of [`src/services/stringCompletions.ts@getStringLiteralTypes` ](https://github.com/microsoft/TypeScript/blob/bd9e711f95cc8734725fbbd7786d6c53a59bcac2/src/services/stringCompletions.ts#L552-L557). Pending on the decision about the best placement of this code (see [here](https://github.com/microsoft/TypeScript/pull/56182/files#r1368390031)), I plan to either deduplicate this or move this whole thing entirely back to services